### PR TITLE
[TRUNK-12396] update cli to work with the new apis

### PR DESCRIPTION
**Pull request by federicoweber**
This updates the quarantine flow to work with the new APIs

### CODE REVIEW

File: cli/src/runner.rs [hunk 1]
```diff
@@ -179,54 +202,66 @@ pub async fn run_quarantine(
...
-            log::error!("{} -> {} [QUARANTINED]", failure.parent_name, failure.name);
+        if quarantined.contains(&failure.id.as_ref().unwrap()) {
+            log::error!(
+                "{} -> {} [QUARANTINED] (id: {:?})",
+                failure.parent_name,
+                failure.name,
+                failure.id.clone().unwrap()
+            );
+
+            quarantine_results.quarantine_results.push(failure.clone());
```
Comment (dfrankland) [inline cli/src/runner.rs:242]: We end up just clone things from `RunResult` and nothing else needs access to it, so `run_quarantine` should accept a value like `run_result: RunResult` in the arguments

File: cli/src/runner.rs [hunk 2]
```diff
@@ -179,54 +202,66 @@ pub async fn run_quarantine(
...
+                failure.id.clone().unwrap()
+            );
         }
     }
+    quarantine_results.group_is_quarantined = run_result.failures.iter().all(|failure| {
+        quarantine_results
+            .quarantine_results
+            .iter()
+            .any(|quarantined_failure| quarantined_failure.id == failure.id)
+    });
```
Comment (dfrankland) [inline cli/src/runner.rs:257]: I think this logic is hard to follow and needlessly clones a lot of things. From what I can see this boils down very simply to the following:
```suggestion
    let total_failures = run_result.failures.len();
    quarantine_results.quarantine_results = run_result
        .failures
        .into_iter()
        .filter_map(|failure| {
            let quarantine_failure = quarantined.contains(&failure.id);
            log::error!(
                "{} -> {}{}(id: {})",
                failure.parent_name,
                failure.name,
                if quarantine_failure {
                    " [QUARANTINED] "
                } else {
                    " "
                },
                failure.id
            );
            if quarantine_failure {
                Some(failure)
            } else {
                None
            }
        })
        .collect();
    quarantine_results.group_is_quarantined =
        quarantine_results.quarantine_results.len() == total_failures;
```

Using `into_iter` relies on the change to accepting `run_result` as a value and not a reference. The reason for using `into_iter` instead of `iter` is to prevent the need for `clone`.

File: cli/src/runner.rs [hunk 3]
```diff
@@ -179,54 +202,66 @@ pub async fn run_quarantine(
...
-    let quarantine_results = if run_result.failures.is_empty() {
-        QuarantineBulkTestStatus {
-            group_is_quarantined: false,
-            quarantine_results: Vec::new(),
-        }
-    } else {
+    let mut quarantine_results = QuarantineBulkTestStatus {
+        group_is_quarantined: false,
+        quarantine_results: Vec::new(),
+    };
```
Comment (dfrankland) [inline cli/src/runner.rs:208]: nit: can we define this below, where we use/mutate it?

nit: Also, we should add a `#[derive(Default)]` to `QuarantineBulkTestStatus` that way we can simply do the following

```suggestion
    let mut quarantine_results = QuarantineBulkTestStatus::default();
```

File: cli/src/runner.rs [hunk 4]
```diff
@@ -179,54 +202,66 @@ pub async fn run_quarantine(
...
+                is_preview_mode: false,
+                quarantined_tests: Vec::new(),
             }
         })
+    } else {
+        log::info!("No failed tests to quarantine");
+        QuarantineConfig {
+            is_preview_mode: false,
+            quarantined_tests: Vec::new(),
+        }
```
Comment (dfrankland) [inline cli/src/runner.rs:228]: nit: similarly here we should use `#[derive(Default)]` on `QuarantineConfig` so we don't have to repeat ourselves and we can refactor this code to be much more simple:

```suggestion
        if let Err(ref err) = result {
            log::error!("Failed to get quarantine results: {:?}", err);
        }

        result.unwrap_or_default()
    } else {
        log::info!("No failed tests to quarantine");
        QuarantineConfig::default()
```

File: cli/src/runner.rs [hunk 5]
```diff
@@ -179,54 +202,66 @@ pub async fn run_quarantine(
...
+                "{} -> {} [QUARANTINED] (id: {:?})",
+                failure.parent_name,
+                failure.name,
+                failure.id.clone().unwrap()
+            );
+
+            quarantine_results.quarantine_results.push(failure.clone());
         } else {
-            log::error!("{} -> {}", failure.parent_name, failure.name);
+            log::error!(
```
Comment (dfrankland) [inline cli/src/runner.rs:244]: an aside: seems odd to me to make these error logs 🤷

File: cli/src/runner.rs [hunk 6]
```diff
@@ -179,54 +202,66 @@ pub async fn run_quarantine(
...
-        .iter()
-        .map(|qr| &qr.name)
-        .collect();
-
+    // quarantine the failed tests
+    let quarantined: HashSet<_> = quarantine_config.quarantined_tests.iter().collect();
     for failure in &run_result.failures {
-        if quarantined_names.contains(&failure.name) {
-            log::error!("{} -> {} [QUARANTINED]", failure.parent_name, failure.name);
+        if quarantined.contains(&failure.id.as_ref().unwrap()) {
```
Comment (zaycev) [inline cli/src/runner.rs:234]: Please don't use unwrap. If you need to get value from option or result, use `expect("Something unexpected happened")`.

File: cli/src/runner.rs [hunk 7]
```diff
@@ -102,7 +103,25 @@ pub fn build_filesets(
     Ok((file_sets, file_counter))
 }
 
+pub fn get_run_info_id(test: &Test, org_slug: &str, repo: &BundleRepo) -> String {
```
Comment (zaycev) [inline cli/src/runner.rs:106]: Please add test.

File: cli/src/runner.rs (hunk 5 above)
Comment (zaycev) [reply cli/src/runner.rs:244]: I agree. How would this work if we have someone with 100k tests failed due to some missconfiguration. I think this should be controller by flag if we even need this.

**Comment by zaycev** [comment]
I would highly recommend to use `log::error` only to log errors. Test failure is not an error, it's expected output.

File: cli/src/runner.rs (hunk 7 above)
Comment (zaycev) [reply cli/src/runner.rs:106]: Also, there are a lot of unnecessary allocations for each ID. Consider re-using buffer which holds input data:
```rust
pub fn get_run_info_id(buf: &mut Vec<u8>, test: &Test, org_slug: &str, repo: &BundleRepo) -> String {
    buf.clear();
    let mut buf = Vec<u8>::new();
    buf.extend_from_slice(org_slug.as_bytes());
    buf.push(b'/');
    buf.extend_from_slice(repo_full_name.as_bytes());
    buf.push(b'/');
    buf.extend_from_slice(test.file.as_deref().unwrap_or("").as_bytes());
    buf.push(b'#');
    buf.extend_from_slice(test.class_name.as_deref().unwrap_or("").as_bytes());
    buf.push(b'#');
    buf.extend_from_slice(test.parent_name.as_bytes());
    buf.push(b'#');
    buf.extend_from_slice(test.name.as_bytes());
    buf.push(b'#');
    buf.extend_from_slice("JUNIT_TESTCASE".as_bytes());
    uuid::Uuid::new_v5(&uuid::Uuid::NAMESPACE_URL, &buf[..]).to_string()
}
```

You can wrap this into `TestIdBuilder` type of you want, which you can create before traversing tests:

```
            let mut id_builder = TestIdBuilder::default();
            for suite in junitxml.suites {
            ...
                      let mut test = Test {
                            parent_name: parent_name.clone(),
                            name: name.clone(),
                            class_name: case.classname.clone(),
                            file: case.file.clone(),
                            id: Some(id_builder.build_test_id(&test, org_slug, repo)),
                        };
```

```

File: cli/src/runner.rs [hunk 8]
```diff
@@ -157,12 +176,16 @@ pub async fn extract_failed_tests(
                     if failure {
                         let name = case.original_name;
                         log::debug!("Test failed: {} -> {}", parent_name, name);
-                        failures.push(Test {
+                        let mut test = Test {
                             parent_name: parent_name.clone(),
```
Comment (zaycev) [inline cli/src/runner.rs:180]: Why do we need to clone these fields if you are already moving cases out of `suite.cases` ?

File: cli/src/runner.rs [hunk 9]
```diff
@@ -157,12 +176,16 @@ pub async fn extract_failed_tests(
...
                             name: name.clone(),
                             class_name: case.classname.clone(),
                             file: case.file.clone(),
-                        });
+                            id: None,
```
Comment (zaycev) [inline cli/src/runner.rs:184]: id: Some(...),

File: cli/src/runner.rs [hunk 10]
```diff
@@ -157,12 +176,16 @@ pub async fn extract_failed_tests(
                     if failure {
                         let name = case.original_name;
                         log::debug!("Test failed: {} -> {}", parent_name, name);
-                        failures.push(Test {
+                        let mut test = Test {
```
Comment (zaycev) [inline cli/src/runner.rs:179]: Does not have to be mut.

File: cli/src/runner.rs (hunk 9 above)
Comment (zaycev) [reply cli/src/runner.rs:184]: Also, consider just adding constructor for `Test` type.

File: cli/src/runner.rs (hunk 5 above)
Comment (federicoweber) [reply cli/src/runner.rs:244]: I hear you, but I think it is much easier to parse the log

Comment (federicoweber) [reply cli/src/runner.rs:244]: Going to move it to info as both you and Vladimir had the same comment

File: junit.xml
Comment (dfrankland) [inline junit.xml:1]: accidental inclusion?

File: cli/src/types.rs [hunk 11]
```diff
@@ -49,28 +47,23 @@ pub struct Test {
...
-    pub quarantine_results: Vec<QuarantineResult>,
+    pub quarantine_results: Vec<Test>,
+}
+
+#[derive(Debug, Serialize, Clone, Deserialize)]
+pub struct QuarantineConfig {
+    #[serde(rename = "isPreview")]
+    pub is_preview_mode: bool,
+    #[serde(rename = "testIds")]
+    pub quarantined_tests: Vec<String>,
```
[the remark this answers is not in the document]
Comment (dfrankland) [reply cli/src/types.rs:98]: As a heads up, you can still make `quarantined_tests` a `HashSet` so you don't need to convert it to one later

File: cli/src/runner.rs [hunk 12]
```diff
@@ -172,61 +178,65 @@ pub async fn extract_failed_tests(
...
     };
 
-    // Print out failed tests and their quarantine status
-    let quarantined_names: HashSet<_> = quarantine_results
-        .quarantine_results
-        .iter()
-        .map(|qr| &qr.name)
+    // quarantine the failed tests
+    let mut quarantine_results = QuarantineBulkTestStatus::default();
+    let quarantined: HashSet<_> = quarantine_config.quarantined_tests.iter().collect();
```
Comment (dfrankland) [inline cli/src/runner.rs:206]: if you define `quarantined_tests` as a `HashSet` you won't have to convert it to one here

**Review by dfrankland — APPROVED** [review]

File: cli/src/runner.rs (hunk 12 above)
Comment (federicoweber) [reply cli/src/runner.rs:206]: ah shoot forgot abut that
